Tucker airtable data pipeline

jtuckerk opened the pull request:
Creating a airtable table syncing pipeline.
So far this is only set up for events

todo:

- [x] initiatives
- [x] roles
- [x] more testing

adamfortuna committed:
Updated README and app for issues that came up during onboarding

adamfortuna committed:
Moved declaration of donation_link_requests db to before its used

adamfortuna committed:
Setting up the API/python side to support dynamic databases and add a bootstrap script to create/populate databases.

adamfortuna committed:
Setting up the API/python side to support dynamic databases and add a bootstrap script to create/populate databases.

adamfortuna committed:
Update bootstrap script to insert initiatives, volunteer events and voluneteer openings

adamfortuna committed:
Update bootstrap script to insert initiatives, volunteer events and voluneteer openings

adamfortuna committed:
Update the API requests to use Pydantic models for return types

adamfortuna committed:
Update the API requests to use Pydantic models for return types

adamfortuna committed:
Allow volunteer roles to return data via FastAPI and pydantic models

adamfortuna committed:
Allow volunteer roles to return data via FastAPI and pydantic models

adamfortuna committed:
Adding in new tests for sqlalchemy

adamfortuna committed:
Adding in new tests for sqlalchemy

adamfortuna committed:
Fix VolunteerRole ID lookup API endpoint filtering

adamfortuna committed:
Fix VolunteerRole ID lookup API endpoint filtering

adamfortuna committed:
Added volunteer events API endpoints

adamfortuna committed:
Added volunteer events API endpoints

adamfortuna committed:
Added initiative model, schema and endpoint - without related models

adamfortuna committed:
Added initiative model, schema and endpoint - without related models

adamfortuna committed:
Updated config for generators

adamfortuna committed:
Updated config for generators

adamfortuna committed:
Added initiaitive -> role relationship and tests

adamfortuna committed:
Added initiaitive -> role relationship and tests

adamfortuna committed:
Uncomment db creation script

adamfortuna committed:
Uncomment db creation script

adamfortuna committed:
Limit sqlalchemy imports to only those that are used

adamfortuna committed:
Limit sqlalchemy imports to only those that are used

adamfortuna committed:
Commenting out UUID columns for now since MySQL doesn't have those

adamfortuna committed:
Commenting out UUID columns for now since MySQL doesn't have those

Calvinxc1 committed:
Added postgres docker scaffolding

Calvinxc1 committed:
Added postgres docker scaffolding

Calvinxc1 committed:
Removed jupyter cache dir

Calvinxc1 committed:
Removed jupyter cache dir

jtuckerk committed:
Single DB config & psql boostrap setup

Reducing the config file and the bootstrap logic to expect only a
  single database. introducing the second database was only intended
  to be temporary to ensure no pii was accidentally shared.

  setting up the bootstrap script to match and starting the transition
  to psql code

jtuckerk committed:
in progress

jtuckerk committed:
Fix filter argument error - api with local db working

jtuckerk committed:
Add events relationships to initiatives.

jtuckerk committed:
Adding donation link request model + minor fixes

jtuckerk committed:
Fix up tests

jtuckerk committed:
add end to end post link request test

jtuckerk committed:
remove base __init__ doesnt seem needed and messes with mypy

jtuckerk committed:
fix linting warnings

adamfortuna committed:
Add additional test to make sure initiatives relationships are scoped

adamfortuna committed:
Adding test for verifying relationships are loading right

adamfortuna committed:
Move test data into test directory

adamfortuna committed:
Changed config example to match current schema used

Calvinxc1 committed:
merge so I can progress

Calvinxc1 committed:
Added docker-compose for postgres, with test data

adamfortuna committed:
Undoing the merge commit. Checked with Jason and looks like the Docker work is in the commit after this one.

This reverts commit 2e98b03395f49bf495b9a20832f619e8d7d6c909, reversing
changes made to 653b17c1e806fd82b81a33ba9886de29cf020883.

adamfortuna committed:
Switched from using postgres data files to use postgres schema dumps

adamfortuna committed:
WIP for the app docker setup:

adamfortuna committed:
Client application working with Docker

adamfortuna committed:
Uncommented the docker db setup since it's working

Calvinxc1 committed:
Working build, missing wait-for-it in api

adamfortuna committed:
API, tests and client running together. Not able to verify DB connection yet

Calvinxc1 committed:
Force rebuild of node_modules properly

adamfortuna committed:
Change node modules folder volume

adamfortuna committed:
Docker compose up from scratch with a single command

adamfortuna committed:
Getting the front-end working end to end

adamfortuna committed:
Updated the readme for docker updates

adamfortuna committed:
Move python config.yml into the repo since we can use it on production

adamfortuna committed:
Fixed existing tests and updated db to include the uuid columns

adamfortuna committed:
Fixing tests for initiatives relationships to events and roles

adamfortuna committed:
Added personal donation link requests

adamfortuna committed:
Merged master back into this branch

adamfortuna committed:
Clean out node_modules before build of the client

adamfortuna committed:
Fix donation email reference in api

adamfortuna committed:
Fix API for donation requests

jtuckerk committed:
Add airtable event syncing & edit event db model

jtuckerk committed:
airtable sync and response converter initial tests.

jtuckerk committed:
Add event sync call to api (temporary?)

jtuckerk committed:
Add Role Syncing logic and minimal testing for role conversion

jtuckerk committed:
Add airtable initiative pipeline and minimal tests

jtuckerk committed:
Update frontend schema names

adamfortuna committed:
Map personal donation table as part of bootstrapping

adamfortuna committed:
Setup the testing environment as its own docker container based on the API container

adamfortuna committed:
Removing deprecated data_source object in favor of sqlalchemy

adamfortuna committed:
Revert from master back to develop

adamfortuna commented on api/api.py:
```diff
@@ -37,26 +38,45 @@ def get_all_volunteer_roles(db: Session = Depends(get_db)) -> List[VolunteerRole
     DonationEmail.insert(donationEmail)
     return donationEmail
+
+def SyncRunner(db, sync_fn) -> str:
```
What would you think about moving from here down to a different file? I saw @trippersham did something like that in the auth branch, ( https://github.com/HF-RapidResponse/volunteer-portal/blob/auth_poc/api/api.py#L11 ) which I think would be a good way to add additional routes without putting everything in api.py.

jtuckerk replied:
done!

adamfortuna commented on api/db_sync_pipelines/pipelines.py:
```diff
@@ -0,0 +1,34 @@
+
+def RunEventsSync(db):
+  table_name = "[Events]"
+  table_key = 'appXC8UaNIuMXpnGB'
```
Do you think we could split these out into a settings variable? It'd be nice to be able to test this using a copy of the Airtable without having access to production.

jtuckerk replied:
I do think it's something we should look into, but maybe further down the road.
I'm not sure I want to spin up a test version of the airtable. I don't see much benefit since any changes to the prod tables wouldn't be caught by those tests. (the airtable team is looking into how to prevent changes to the tables and columns we care about.)

It doesn't look like there are very fine grained permissions within airtable so giving permissions to a view or even a "synced" version of the table would open up all of airtable ~ish. ish: If we can change those table keys ^ we could keep the new ones private (should have kept them private to start) and only expose the global key (which we'd have to verify doesn't grant access to downloading individual table keys, or anything else we dont want)

option I like the most would be while syncing, dump a subset the json response to a cloud storage file and use something like the FakeAirtableLoader ive used in the tests to serve as a dev/test replacement for airtable

adamfortuna commented on api/models/volunteer_role.py:
```diff
@@ -12,31 +12,31 @@
     overview = Column('job_overview', Text)
     benefits = Column('what_youll_learn', Text)
-    responsibilites = Column('responsibilities_and_duties', Text)
+    responsibilities = Column('responsibilities_and_duties', Text)
```
lol, good catch!

jtuckerk replied:
:)

adamfortuna commented on api/models/initiative.py:
```diff
@@ -12,58 +12,61 @@
     role_ids = Column('roles', ARRAY(String))
     event_ids = Column('events', ARRAY(String))
+    airtable_last_modified = Column('airtable_last_modified', DateTime, nullable=False)
+    db_last_modified = Column('db_last_modified', DateTime, nullable=False)
```
This is 100% personal preference, but I always call this column `updated_at`. It looks like there's a way to automatically update this column too: https://stackoverflow.com/questions/13370317/sqlalchemy-default-datetime

```
from sqlalchemy.sql import func
updated_at = Column('updated_at', DateTime, onupdate=func.now())
```

Could be one wy to slim down a little code.
With that, SQLAlchemy will automatically update that column's value whenever it changes.

jtuckerk replied:
oh yeah much better. thanks!

adamfortuna commented on api/tests/db_sync_pipelines/events_example_response.json:
```diff
@@ -0,0 +1,223 @@
+      "Last Modified": "2020-11-20T16:42:54.000Z",
+      "Created by": {
+        "id": "usrg7iGBoicYG7cfL",
+        "email": "tom@tomjeffreyinnovations.com",
```
Might want to remove these emails, and change some names to use `@example.com` if they're real ones.

jtuckerk replied:
ah good catch!

jtuckerk replied:
I'll rebase this and force push the changes before we merge this just to remove that from the commit graph, but wanted to leave you with the diffs for now.

adamfortuna commented on api/tests/db_sync_pipelines/events_example_response.json:
```diff
@@ -0,0 +1,223 @@
+      "Last Modified": "2020-11-20T16:42:54.000Z",
+      "Created by": {
+        "id": "usrg7iGBoicYG7cfL",
+        "email": "tom@tomjeffreyinnovations.com",
```
Might want to remove these emails, and change some names to use `@example.com` if they're real ones.

jtuckerk replied:
thanks!

adamfortuna reviewed (changes requested):
Few minor recommendations, but looks good to me!

adamfortuna committed:
Added donation request API endpoint

jtuckerk committed:
last_modified->updated at. split out sync routes. pii clean.

jtuckerk committed:
remove debug raise

jtuckerk committed:
Add sync test and fix/finish sync pipeline

adamfortuna committed:
Switch donation email to use email pydantic type

adamfortuna committed:
Merge branch 'develop' into adam-multiple-environments

adamfortuna committed:
Mering develop/client back in

adamfortuna committed:
Update GitHub tests to run python using new bootstrap setup

adamfortuna committed:
Switching to a single CI file that uses docker compose for github testing

adamfortuna committed:
Fix client CI command for GH workflow

adamfortuna committed:
Fixing the client docker name to the actual name and not the makefile alias

adamfortuna committed:
Install all depenencies with NPM for CI purposes

adamfortuna committed:
Exit client tests in CI after running

adamfortuna committed:
Fix spacing in github workflow file to be consistent

jtuckerk committed:
Put back client code to latest develop changes. (#80)

* Put back client code to latest develop changes.

* add missing paren

* fix initiatives path

jtuckerk committed:
Add in db schema changes

jtuckerk committed:
Merge branch 'adam-multiple-environments' into tucker-airtable-data-pipeline

jtuckerk committed:
Merge branch 'develop' into tucker-airtable-data-pipeline

jtuckerk committed:
Merge branch 'develop' into tucker-airtable-data-pipeline

jtuckerk committed:
Adding more tests and db changes.

Enforcing that fields are not null in the DB and setting up tests to
  make sure that lines up with the API definitions in api/schema

jtuckerk committed:
fix flakey test.

jtuckerk committed:
Actually fix flakes. Added db cleanup logic.

trippersham commented on api/db_sync_pipelines/airtable_event_converter.py:
```diff
@@ -0,0 +1,26 @@
+  }
+
+  custom_transforms = {
+      'start_datetime': ParseTimestamp,
```
I like that these are defined functions (as opposed to the lambdas I was using previously). I feel like it will be good to have a library of these common "custom" transforms when we have other data sources in the pilot even if we need to support truly one-off transforms in the future.

jtuckerk replied:
Yeah, I thought about pulling these out fully to a separate utils file, but since theyre so simple and the only real meat of the function is an airtable date format,  I left them in an airtable file.

jtuckerk replied:
I'm leaning towards leaving it for now but if you feel strongly about breaking it out now, let me know

jtuckerk replied:
moved these to the api_response_converter file since the now timestamp is generic and airtable uses the ISO standard  (also added this and other comments to a review to avoid individual notifications for each comment I make, sorry if it formats weird and any previous annoyance)

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+                    hard_delete=False):
+
+    db_model = response_converter.GetDBModel()
+    all_records = list(GetAndConvertAirtableRecords(
```
GetAndConvertAirtableRecords is a generator but you immediately convert it to a list to then use in a loop. Unless I misunderstand generators (very possible) you're losing the performance benefits of using the generator in the first place; why not just pass the generator straight into the loop?

trippersham replied:
Digging through the calls: why are you using a generator? There's only one network call and you get a list of all of the airtable rows (I'm assuming the memory usage of that list is not substantially different than the converted rows).

jtuckerk replied:
I probably put that in for debugging and accidentally left it. It's now removed.
I was just using a generator more for style and simplicity to avoid declaring a list and appending. Also possible I borrowed code from something that did some paging.
I'm now using list comprehension which may be bad style, idk.

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+
+
+def GetExistingRecordIds(db, db_model, hard_delete=False):
+    query = db.query(db_model.external_id)
```
This creates a hard requirement that all external data sources need to have a persistent unique ID associated with each row. Not sure if that's a good or bad requirement, but something we should keep in mind for the future.

jtuckerk replied:
For the data centralization work I was doing I was creating a separate syncing pipeline for each platform. definitely ended up being a lot of code reused, but then you dont have to worry as much about things like that. Since this is just the airtable syncer and it seems like we're moving away from the possibility of using more external platforms in the near to mid term I'm not too concerned
good to keep in mind for sure tho

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+
+def GetLatestSyncRun(db, db_model):
+    query = db.query(db_model.updated_at).order_by(db_model.updated_at.desc())
+    result = query.first()
```
This assumes those records can only be modified by the sync pipeline. We don't have a way to edit these records from the UI today but we probably will in the future (even if we don't allow you to edit/override this data, we may have other fields like "users_with_this_role" that are updated on our side). It may be safer/more future-proofed to keep a separate updated column that only pertains to the sync pipeline.

jtuckerk replied:
hmmm good point.
How do you feel about leaving it as is and just moving off airtable once we have a portal UI for updating them? The idea of having 2 sources for 1 db doesn't seem fun. If we do leave this field to be updated by the UI or the pipeline then it would effectively give the UI priority since anything changed with the UI later would cause earlier Airtable changes to be ignored. which I think would be fine behavior, tho again, not excited about dealing with figuring that out completely. 20% chance I'm ignoring other good reasons to change it and am just coming up with reasons not to do any more work.

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+    for record in all_records:
+        if record.external_id not in existing_record_ids:
+            new_records.append(record)
+        elif NeedsUpdate(record, latest_sync_run):
```
Smart, I like this

jtuckerk replied:
May be overkill for some of these smaller tables, but if we ever decide to pull all the volunteer records in, this gives a big speedup.

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+        elif NeedsUpdate(record, latest_sync_run):
+            updated_records.append(record)
+
+    airtable_records = set([record.external_id for record in all_records])
```
slight suggestion to change these variable names to "_ids" so new folks don't think the contents of "new_records" and "deleted_records" are the same types

jtuckerk replied:
good call

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+        for local_record in local_records:
+            # Feels like there should be a better way to do this.
+            db.query(db_model).filter(
+                db_model.external_id == local_record.external_id).delete()
```
If/when we have other data stored on those records that doesn't come from this external source, this will be overzealous. Fine for now, but we should probably do an upsert as opposed to a delete and replace.

jtuckerk replied:
yeah,  I didnt like this. I banged my head on it for a while and didnt feel like I was getting the type of solution I thought should exist. I may not have searched "upsert" hmmm. but either way I'm still a bit wary of having multiple sources for a single record

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+        raise
+
+
+def UpdateRecordsAsDeleted(to_delete, db, db_model):
```
Love a good soft delete

trippersham commented on api/db_sync_pipelines/airtable_sync.py:
```diff
@@ -0,0 +1,166 @@
+  return datetime.strptime(timestamp, AIRTABLE_DATETIME_FORMAT)
+
+
+class AirtableLoader:
```
What's the rationale for having these functions live outside of the class? Making testing easier so we don't need to use a mock?

jtuckerk replied:
at some point in my career someone who cares a lot more about these sorts of things than I do pointed out that if a function doesnt operate on an instance's state it shouldn't be defined within the class. It avoids having to use self in the definition and the call which i guess is cleaner.
~time passes and other comments got me thinking~
I've made these part of the APIResponseConverter file, but still not a part of the class for the same reason

trippersham commented on api/models/initiative.py:
```diff
@@ -2,68 +2,72 @@
+from sqlalchemy.sql import func
 from uuid import uuid4
 
 class Initiative(Base):
```
What do you think about creating a subclass of Base that handles things like "external_id", "is_deleted", and "updated_at" (esp. depending on your thoughts on my comment on GetLatestSyncRun)? Will make it harder to accidentally forget those fields that are primarily used by the sync pipeline.

jtuckerk replied:
ooh yeah

jtuckerk replied:
done

jtuckerk committed:
PR Review changes. Class Organization mostly

jtuckerk reviewed (commented):
great feedback thank you!
